### PR TITLE
Support `ServiceAccount` signing key secret rotation for shoot clusters

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area security
/kind enhancement

**What this PR does / why we need it**:
This PR introduces the possibility to rotate the service account signing key secrets for shoot clusters. This feature is behind a newly introduced `ShootSARotation` feature gate, mainly because the extensions need to revendor `gardener/gardener` in order to support it.

The process is very similar to how shoot CA rotation works:

- There are two phases: In phase 1, a new service account signing key is generated and used for issuing new tokens. The old signing key is still used for verification. In phase 2, the old signing key is dropped.
- In phase 1, all shoot access secrets must be renewed (new tokens signed by new key must be generated).
- Also, all static `ServiceAccount` token secrets must be rotated (new secrets with tokens signed by new key must be generated).
- In phase 2, all old static `ServiceAccount` token secrets must get deleted (they are now invalid since the old signing key is no longer accepted).

This PR does intentionally not include documentation since I would like to wait for #5939 to be merged. Respective documentation will be submitted accordingly afterwards.

**Release note**:

```feature operator
A new alpha `ShootSARotation` feature gate (disabled by default) has been introduced which allows the rotation of the service account signing key secrets for shoot clusters.
```

### CODE REVIEW

File: docs/deployment/feature_gates.md [hunk 1]
```diff
@@ -39,6 +39,7 @@ The following tables are a summary of the feature gates that you can set on diff
 | ForceRestore                                 | `false` | `Alpha` | `1.39` |        |
 | DisableDNSProviderManagement                 | `false` | `Alpha` | `1.41` |        |
 | ShootCARotation                              | `false` | `Alpha` | `1.42` |        |
+| ShootSARotation                              | `false` | `Alpha` | `1.48` |        |
```
Comment (ialidzhikov) [inline docs/deployment/feature_gates.md:43]: > This PR does intentionally not include documentation since I would like to wait for https://github.com/gardener/gardener/pull/5939 to be merged. Respective documentation will be submitted accordingly afterwards.

As #5939 is now merged do you plan to add docs as part of this PR or will it be done in a separate PR?

File: pkg/apis/core/helper/helpers.go [hunk 2]
```diff
@@ -371,6 +371,17 @@ func GetShootCARotationPhase(credentials *core.ShootCredentials) core.ShootCrede
 	return ""
 }
 
+// GetShootServiceAccountKeyRotationPhase returns the specified shoot service account key rotation phase or an empty
+// string.
+func GetShootServiceAccountKeyRotationPhase(credentials *core.ShootCredentials) core.ShootCredentialsRotationPhase {
+	if credentials != nil &&
+		credentials.Rotation != nil &&
+		credentials.Rotation.ServiceAccountKey != nil {
```
Comment (ialidzhikov) [inline pkg/apis/core/helper/helpers.go]: Nit: for consistency reasons with the v1beta1 helper func

```suggestion
	if credentials != nil && credentials.Rotation != nil && credentials.Rotation.ServiceAccountKey != nil 
```

File: pkg/apis/core/validation/shoot.go [hunk 3]
```diff
@@ -1747,29 +1747,48 @@ func validateShootOperation(operation string, shoot *core.Shoot, fldPath *field.
...
+		if helper.GetShootServiceAccountKeyRotationPhase(shoot.Status.Credentials) != core.RotationPrepared {
+			allErrs = append(allErrs, field.Forbidden(fldPath, "cannot complete service account key rotation if .status.credentials.rotation.serviceAccountKey.phase is not 'Prepared'"))
+		}
 	}
 
 	return allErrs
 }
 
-func isShootReadyForCARotationStart(lastOperation *core.LastOperation) bool {
+func isShootReadyForRotationStart(lastOperation *core.LastOperation) bool {
```
Comment (ialidzhikov) [inline pkg/apis/core/validation/shoot.go:1791]: @acumino raised a good point - do we forbid for example a SA rotation start when the Shoot is already in phase 1 of CA rotation. I assume it is not supported to run both in parallel, right? Do we have a validation about this case?

File: pkg/registry/core/shoot/strategy_test.go [hunk 4]
```diff
@@ -455,6 +468,66 @@ var _ = Describe("Strategy", func() {
 				shootregistry.Strategy.PrepareForUpdate(context.TODO(), newShoot, oldShoot)
 				Expect(newShoot.Generation).To(Equal(oldShoot.Generation))
 			})
+
+			It("should increase when the rotate-serviceaccount-key-start annotation gets set and feature gate is enabled", func() {
+				defer test.WithFeatureGate(utilfeature.DefaultFeatureGate, features.ShootSARotation, true)()
```
Comment (ialidzhikov) [inline pkg/registry/core/shoot/strategy_test.go]: Nit: does it make sense to use a table test? It looks like the only change between the ITs is the feature gate value and the operation annotation value on the new Shoot.

**Review by ialidzhikov — COMMENTED** [review]
We started looking with @acumino and @shafeeqes . The PR is structured well with commits that are easier to review. For now I am adding few inline nits/questions.

File: pkg/apis/core/validation/shoot.go (hunk 3 above)
Comment (rfranzke) [reply pkg/apis/core/validation/shoot.go:1791]: Why would it not be supported?

File: docs/deployment/feature_gates.md (hunk 1 above)
Comment (rfranzke) [reply docs/deployment/feature_gates.md:43]: Yes, now added with 5e5a2bf.

File: pkg/utils/kubernetes/health/deployment.go [hunk 5]
```diff
@@ -108,3 +114,35 @@ func IsDeploymentProgressing(deployment *appsv1.Deployment) (bool, string) {
 
 	return false, "Deployment is fully rolled out"
 }
+
+// IsDeploymentUpdated returns a function which can be used for retry.Until. It checks if the deployment is fully
+// updated, i.e. if it is no longer progressing, healthy, and has the exact number of desired replicas.
+func IsDeploymentUpdated(reader client.Reader, deployment *appsv1.Deployment) func(context.Context) (bool, error) {
```
Comment (ialidzhikov) [inline pkg/utils/kubernetes/health/deployment.go:121]: It is sad that this case is not reflected in the Deployment status we have to introduce similar func that explicitly checks for terminating Pods. Should we open an upstream issue that it is not possible to determine this by only using the Deployment status?

File: pkg/utils/kubernetes/health/deployment.go [hunk 6]
```diff
@@ -108,3 +114,35 @@ func IsDeploymentProgressing(deployment *appsv1.Deployment) (bool, string) {
...
+			return retry.MinorError(errors.New(reason))
+		}
+
+		// If Deployment is no longer progressing then check if it is healthy.
+		if err := CheckDeployment(deployment); err != nil {
+			return retry.MinorError(err)
+		}
+
+		// Now there might be still pods in the system belonging to an older ReplicaSet of the Deployment.
+		podList := &corev1.PodList{}
```
Comment (ialidzhikov) [inline pkg/utils/kubernetes/health/deployment.go]: Is it possible to use `PartialObjectMetadataList` to prevent fetching the spec and status of the Pods? At the end we are only interested in the Pods count.

File: docs/usage/shoot_credentials_rotation.md [hunk 7]
```diff
@@ -170,9 +171,50 @@ It is currently **not** rotated automatically and there is no way to trigger it
 
 ## `ServiceAccount` Token Signing Key
 
-This key is used to sign the tokens for `ServiceAccount`s.
-It is currently **not** rotated automatically and there is no way to trigger it manually.
-Note that there is the option to pass such key to Gardener, please see [this document](shoot_serviceaccounts.md#signing-key-secret) for more details.
+Gardener generates a key which is used to sign the tokens for [`ServiceAccount`s](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/).
```
Comment (ialidzhikov) [inline docs/usage/shoot_credentials_rotation.md:175]: `docs/usage/shoot_serviceaccounts.md` needs update - it says:
```
Note that rotation of this key is not yet supported, hence usage is not recommended.
```

File: docs/usage/shoot_credentials_rotation.md [hunk 8]
```diff
@@ -170,9 +171,50 @@ It is currently **not** rotated automatically and there is no way to trigger it
...
+Those tokens are typically used by workload `Pod`s running inside the cluster in order to authenticate themselves with the `kube-apiserver`.
+This also includes system components running in the `kube-system` namespace.
+
+The token signing key has no expiration date.
+Since it might require adaptation for the consumers of the `Shoot`, there is no automatic rotation and **it is the responsibility of the end-user to regularly rotate the signing key.**
```
Comment (ialidzhikov) [inline docs/usage/shoot_credentials_rotation.md:180]: Is this signing key the same one that can be configured with the [`kubeAPIServer.serviceAccountConfig.signingKeySecretName.name`](https://github.com/gardener/gardener/blob/cd0fc9f26f6d63d37545b45eb476a2f7858020f2/example/90-shoot.yaml#L184-L185)? If yes, then we could mention that it is the responsibility of the end-user in case they don't configure `kubeAPIServer.serviceAccountConfig.signingKeySecretName.name`. Do you know how the rotation would look like when the Secret is not gardener managed one? We could describe the rotation in that case as well.

File: docs/usage/shoot_credentials_rotation.md [hunk 9]
```diff
@@ -170,9 +171,50 @@ It is currently **not** rotated automatically and there is no way to trigger it
 
 ## `ServiceAccount` Token Signing Key
```
Comment (ialidzhikov) [inline docs/usage/shoot_credentials_rotation.md:173]: There are no integration/e2e tests added as part of this PR for the new feature (which can be acceptable having in mind that it is behind alpha feature gate). Can you create new issue (or new item in umbrella issue) to add such tests?

**Review by ialidzhikov — COMMENTED** [review]
Overall lgtm. Thanks for the docs and the well-structured commits!
I am adding as inline comments the few things that I spotted from my review. I didn't try out the feature locally.

File: pkg/apis/core/validation/shoot.go (hunk 3 above)
Comment (ialidzhikov) [reply pkg/apis/core/validation/shoot.go:1791]: If you say that it is supported, then ok. It was just a question that we raised during our review back then.

File: pkg/utils/kubernetes/health/deployment.go (hunk 6 above)
Comment (rfranzke) [reply pkg/utils/kubernetes/health/deployment.go]: Awesome idea, will do :)

File: pkg/utils/kubernetes/health/deployment.go (hunk 5 above)
Comment (rfranzke) [reply pkg/utils/kubernetes/health/deployment.go:121]: Well, I'm not too sure about this question. I would expect that this is on purpose (`Deployment`'s `Progressing` condition only cares whether the desired `Pod`s were created).

File: docs/usage/shoot_credentials_rotation.md (hunk 8 above)
Comment (rfranzke) [reply docs/usage/shoot_credentials_rotation.md:180]: Currently, no rotation is implemented for user-provided signing keys, and this field gets deprecated with https://github.com/gardener/gardener/pull/5980 and removed eventually. I would not invest anything into it now (also no docs).

File: docs/usage/shoot_credentials_rotation.md (hunk 9 above)
Comment (rfranzke) [reply docs/usage/shoot_credentials_rotation.md:173]: We will add the etcd encryption secret rotation as well and then refactor the existing e2e test which only covers CA rotation today as part of https://github.com/gardener/gardener/issues/3292.
Eventually, we envision having only one e2e test suite for the credentials rotation.

File: docs/usage/shoot_credentials_rotation.md (hunk 8 above)
Comment (ialidzhikov) [reply docs/usage/shoot_credentials_rotation.md:180]: Ok, I didn't notice that it is getting deprecated.

File: docs/usage/shoot_credentials_rotation.md (hunk 7 above)
Comment (rfranzke) [reply docs/usage/shoot_credentials_rotation.md:175]: I think this is still valid since this speaks about the user-provided signing key. Ref https://github.com/gardener/gardener/pull/5968#discussion_r876012279

File: pkg/apis/core/validation/shoot.go (hunk 3 above)
Comment (acumino) [reply pkg/apis/core/validation/shoot.go:1791]: Service account secret has ca.crt. 
So for example when the ca rotation starts and is in phase 1, at this stage will this service account secret contain both ca.crt? Also if at this stage if the user triggers SA rotation and completes it will the new SA token can be used with both CA? Maybe it's very obvious thing that I am not able to get but can you please point to the code which handles these scenarios.

Comment (rfranzke) [reply pkg/apis/core/validation/shoot.go:1791]: > So for example when the ca rotation starts and is in phase 1, at this stage will this service account secret contain both ca.crt?

Yes

> Also if at this stage if the user triggers SA rotation and completes it will the new SA token can be used with both CA? 

CA certificates and SA tokens are not related to each other, so I don't really understand this question.

> Maybe it's very obvious thing that I am not able to get but can you please point to the code which handles these scenarios.

Which code are you looking for/which scenarios?

**Review by ialidzhikov — COMMENTED** [review]
/lgtm

**Comment by rfranzke** [comment]
Rebased to latest `master`

**Review by ialidzhikov — COMMENTED** [review]
/lgtm

**Review by ialidzhikov — COMMENTED** [review]
/lgtm
